tests/prague/eip2935_historical_block_hashes_from_state/test_block_hashes.py

```diff
@@ -0,0 +1,175 @@
+                block_outside_window = i - Spec.BLOCKHASH_OLD_WINDOW - 1
+
+                # Check using BLOCKHASH, must be zero
+                code += Op.SSTORE(
+                    storage.store_next(1), Op.ISZERO(Op.BLOCKHASH(block_outside_window))
+                )
+
+                # Check calling the HISTORY_STORAGE_ADDRESS, must be zero
+                code += Op.MSTORE(0, block_outside_window)
+                code += Op.POP(Op.CALL(Op.GAS, Spec.HISTORY_STORAGE_ADDRESS, 0, 0, 32, 0, 32))
+                code += Op.RETURNDATACOPY(0, 0, 32)
+                code += Op.SSTORE(storage.store_next(1), Op.ISZERO(Op.MLOAD(0)))
+
+            else:
+                # No block outside the window
+                pass
+
+            # Check the oldest block inside the window
+            if i > Spec.BLOCKHASH_OLD_WINDOW:
+                oldest_block = i - Spec.BLOCKHASH_OLD_WINDOW
+            else:
+                oldest_block = 0
+
+            # Check using BLOCKHASH, must be greater than zero
+            code += Op.SSTORE(storage.store_next(1), Op.GT(Op.BLOCKHASH(oldest_block), 0))
+
+            # Check calling the HISTORY_STORAGE_ADDRESS, must be zero
+            code += Op.MSTORE(0, oldest_block)
+            code += Op.POP(Op.CALL(Op.GAS, Spec.HISTORY_STORAGE_ADDRESS, 0, 0, 32, 0, 32))
+            code += Op.RETURNDATACOPY(0, 0, 32)
+            code += Op.SSTORE(storage.store_next(1), Op.ISZERO(Op.MLOAD(0)))
+
+            txs.append(
+                Transaction(
+                    to=current_code_address,
+                    gas_limit=10_000_000,
+                    nonce=next(tx_nonce),
+                )
+            )
+            pre[Address(current_code_address)] = Account(code=code, nonce=1)
+            post[Address(current_code_address)] = Account(storage=storage)
+            current_code_address += 0x100
+        blocks.append(Block(timestamp=i, txs=txs))
+
+    # Add the fork block
+    current_block_number = len(blocks) + 1
+    txs = []
+    # On the block after the fork, BLOCKHASH must return values for the last
+    # Spec.HISTORY_SERVE_WINDOW blocks.
+    # And HISTORY_STORAGE_ADDRESS should be also serve the same values.
+    code = b""
+    storage = Storage()
+    if current_block_number > Spec.HISTORY_SERVE_WINDOW:
+        # We can check that blocks before the window are not available
+        block_outside_window = current_block_number - Spec.HISTORY_SERVE_WINDOW - 1
+
+        # Check using BLOCKHASH, must be zero
+        code += Op.SSTORE(storage.store_next(1), Op.ISZERO(Op.BLOCKHASH(block_outside_window)))
+
+        # Check calling the HISTORY_STORAGE_ADDRESS, must be zero
+        code += Op.MSTORE(0, block_outside_window)
+        code += Op.POP(Op.CALL(Op.GAS, Spec.HISTORY_STORAGE_ADDRESS, 0, 0, 32, 0, 32))
+        code += Op.RETURNDATACOPY(0, 0, 32)
+        code += Op.SSTORE(storage.store_next(1), Op.ISZERO(Op.MLOAD(0)))
```

Comment:
I think we could make this code a reusuable pytest fixture in the future, if we decide to keep it post spec change.

Reply:
Tried my best, but since we are not doing one check per parameterization, I reduced the code generation to a function, and it's a bit more readable.